Sim multidrone

yycho0108 opened the pull request:
Sim now works with multiple drones!

yycho0108 commented:
Currently the topic namespaces and the tf_prefix are sort of mixed, and at some point the package should support namespaces properly (i.e. without the hacky way)

yycho0108 commented:
There's a strange namespacing issue that I discovered recently on the sim configuration. PR should not be merged until that gets resolved.

yycho0108 commented:
Fixed the namespaces issue - technically not the best solution with `tf_prefix` and `ns` mixed together, but works for now and will continue to work unless some strange issue arises.

HALtheWise commented on iarc_sim_3d/launch/sim.launch:
```diff
@@ -1,35 +1,43 @@
+        <!-- single drone cfg -->
+        <include file="$(find iarc_sim_3d)/launch/noisy_odom.launch" />
+        <node pkg="tf" type="static_transform_publisher" name="sim_drone_tf"
+            args="0 0 0 0 0 0 drone/base_link ardrone_base_link 1"/>
```
Do you remember why we added this publisher? My guess is that some stuff in the arbiter explicitly depends on the ardrone_base_link topic, in which case we should fix that and add namespacing support there as well.

HALtheWise commented on iarc_sim_3d/launch/sim.launch:
```diff
@@ -1,35 +1,43 @@
+            args="-0.5 0 0 0 0 0 map drone_2/odom 1"/>
+    </group>
+
+    <group if="false">
```
We should really split into two launch files, sim_single_drone and sim_two_drones, so that we don't have an if(false){} hanging around here, and so we can seamlessly continue to use single-drone simulation when appropriate.

HALtheWise commented on iarc_sim_3d/launch/spawn_drone.launch:
```diff
@@ -49,6 +57,7 @@
     </node>
+    -->
 
     <!-- node name="hector_pose_estimation" pkg="hector_pose_estimation" type="hector_pose_estimation_node" output="screen"/-->
```
I would remove this dead code (or move it to the single-drone launch file if it is still relevant there. We can always dig it out from the git history or last year's codebase if we need it, and having a bunch of commented code in the repo isn't good for code health.

HALtheWise commented on libraries/tum_simulator/cvg_sim_gazebo_plugins/src/quadrotor_simple_controller.cpp:
```diff
@@ -98,12 +98,14 @@ void GazeboQuadrotorSimpleController::Load(physics::ModelPtr _model, sdf::Elemen
     link_name_ = _sdf->GetElement("bodyName")->GetValue()->GetAsString();
-    link = boost::dynamic_pointer_cast<physics::Link>(world->GetEntity(link_name_));
+    //ROS_INFO("link name : %s", link_name_.c_str());
+    //link = boost::dynamic_pointer_cast<physics::Link>(world->GetEntity(link_name_));
```
Nit: Remove debug code

HALtheWise commented on libraries/tum_simulator/cvg_sim_gazebo_plugins/src/quadrotor_simple_controller.cpp:
```diff
@@ -242,13 +244,15 @@ void GazeboQuadrotorSimpleController::VelocityCallback(const geometry_msgs::Twis
 
 void GazeboQuadrotorSimpleController::ImuCallback(const sensor_msgs::ImuConstPtr& imu)
 {
+    //ROS_INFO_THROTTLE(1.0, "IMU");
```
Nit: Remove debug code

HALtheWise commented on libraries/tum_simulator/cvg_sim_gazebo_plugins/src/quadrotor_simple_controller.cpp:
```diff
@@ -242,13 +244,15 @@ void GazeboQuadrotorSimpleController::VelocityCallback(const geometry_msgs::Twis
 
 void GazeboQuadrotorSimpleController::StateCallback(const nav_msgs::OdometryConstPtr& state)
 {
+    //ROS_INFO_THROTTLE(1.0, "STATE");
```
Nit: Remove debug code

HALtheWise commented on libraries/tum_simulator/cvg_sim_gazebo_plugins/src/quadrotor_state_controller.cpp:
```diff
@@ -419,7 +419,7 @@ void GazeboQuadrotorStateController::Update()
 
   navdata.header.stamp = ros::Time::now();
-  navdata.header.frame_id = "ardrone_base_link";
+  navdata.header.frame_id = namespace_ + "base_link"; // TODO : fix - technically tf_prefix
```
Do we support tf_prefix and namespace being different? It is probably OK if we don't, but we should be explicit about that. I'm worried about this causing a lot of confusion for someone in the future.

yycho0108 replied:
we do not currently suppor tf_prefix and namespace being different, and I'm currently slightly lazy to fix it.

HALtheWise commented on libraries/tum_simulator/cvg_sim_gazebo_plugins/urdf/quadrotor_sensors.urdf.xacro:
```diff
@@ -17,8 +17,9 @@ xmlns:xacro="http://ros.org/wiki/xacro"
-        <topicName>/ardrone/imu</topicName>
+        <!-- TODO : update bodyName?? -->
+        <frameId>$(arg tf_prefix)/base_link</frameId>
+        <topicName>$(arg tf_prefix)/imu</topicName>
```
If we are supporting different tf_prefix and namespace, the topic here should probably use the namespace. Likewise below.

HALtheWise commented on libraries/tum_simulator/cvg_sim_gazebo_plugins/urdf/quadrotor_simple_controller.urdf.xacro:
```diff
@@ -1,47 +1,50 @@
+                <navdataTopic>ardrone/navdata</navdataTopic>
+            </plugin>
+        </gazebo>
+    </xacro:macro>
```
Autoformatting this file made reviewing it trickier, we should probably have a standard XML linter we run on the whole codebase at some point in the future. For now, it's fine.

HALtheWise commented on iarc_sim_3d/launch/spawn_drone.launch:
```diff
@@ -21,26 +21,34 @@
+            <param name="child_frame_id" value="$(arg ns)/base_link" /> <!-- maybe? -->
+
+            <param name="tf_prefix" type="string" value="$(arg ns)" />
+            <remap from="tf" to="tf_drone"/>
```
Do we need this whole remap now that the prefix is getting handled differently? I seem to remember that we were publishing to a separate tf topic, then using a different node to read those messages, modify them, and send them along again, and it would be _really_ nice if we didn't need to do that anymore.

yycho0108 replied:
not needed anymore.

yycho0108 commented:
Added 2 configs and removed deprecated/testing code.
the biggest action item is tf_prefix and namespace handling, which I don't think will cause any problems in the near future.
